### PR TITLE
Merge reflection class into the IBMiDb2Dialect

### PR DESCRIPTION
**Pull request by naveen-ram**
The reflection class is now not needed because there is only one
backend, so it was merged with the IBMiDb2Dialect

Fixes: #27

### CODE REVIEW

File: sqlalchemy_ibmi/base.py [hunk 1]
```diff
@@ -677,10 +699,12 @@ class IBMiDb2Dialect(default.DefaultDialect, PyODBCConnector):
...
-        self._reflector = self._reflector_cls(self)
+    """ methods in Db2 on i """
+    def __init__(self, dialect, **kwargs):
+        super().__init__(**kwargs)
+        self.dialect = dialect
+        self.ischema_names = dialect.ischema_names
+        self.identifier_preparer = dialect.identifier_preparer
 
     # reflection: these all defer to an BaseDb2Reflector
     # object which selects between Db2 and AS/400 schemas
```
Comment (kadler) [inline sqlalchemy_ibmi/base.py]: These comments are no longer needed since we don't have any reflection

File: sqlalchemy_ibmi/base.py [hunk 2]
```diff
@@ -730,68 +754,6 @@ def set_isolation_level(self, dbapi_conn, level):
     def get_isolation_level(self, dbapi_conn):
         pass
 
-    def normalize_name(self, name):
-        return self._reflector.normalize_name(name)
-
-    def denormalize_name(self, name):
-        return self._reflector.denormalize_name(name)
```
Comment (kadler) [inline sqlalchemy_ibmi/base.py]: These seem to have been removed completely having been defined in https://github.com/IBM/sqlalchemy-ibmi/pull/34/files?utf8=%E2%9C%93&diff=unified&w=1#diff-3d4ffd35dfc8aebc13a4f5024a07dbc9L56-L69

Are they no longer needed?

File: sqlalchemy_ibmi/base.py [hunk 3]
```diff
@@ -730,68 +754,6 @@ def set_isolation_level(self, dbapi_conn, level):
...
-            connection, table_name, schema=schema, **kw)
-
-    def get_incoming_foreign_keys(
-            self,
-            connection,
-            table_name,
-            schema=None,
-            **kw):
-        return self._reflector.get_incoming_foreign_keys(
-            connection, table_name, schema=schema, **kw)
```
Comment (kadler) [inline sqlalchemy_ibmi/base.py]: Huh, this seems to have been something unique to ibm_db_sa and only implemented for Db2 LUW.

**Comment by kadler** [comment]
Looks like there's some conflicts that need to be resolved.

File: sqlalchemy_ibmi/base.py (hunk 2 above)
Comment (naveen-ram) [reply sqlalchemy_ibmi/base.py]: Found that the implementation they have causes errors when running tests, but the super implementation of those methods works without errors.

File: sqlalchemy_ibmi/base.py (hunk 3 above)
Comment (naveen-ram) [reply sqlalchemy_ibmi/base.py]: Is there a reason this wouldn't work on IBM i ?
https://github.com/IBM/sqlalchemy-ibmi/commit/842a2c7393971676555949b87c4c81e4a60e4dfc

File: sqlalchemy_ibmi/base.py [hunk 4]
```diff
@@ -636,19 +637,41 @@ def get_rowcount(self):
...
+        return value
+
+
 class IBMiDb2Dialect(default.DefaultDialect, PyODBCConnector):
 
     name = 'sqlalchemy_ibmi'
     max_identifier_length = 128
     encoding = 'utf-8'
     default_paramstyle = 'qmark'
     COLSPECS = COLSPECS
```
Comment (kadler) [inline sqlalchemy_ibmi/base.py]: This also looks like it shouldn't have been upper-cased in the class: https://docs.sqlalchemy.org/en/13/core/internals.html

> colspecs
>    A dictionary of TypeEngine classes from sqlalchemy.types mapped to subclasses that are specific to the dialect class. This dictionary is class-level only and is not accessed from the dialect instance itself.

Comment (kadler) [reply sqlalchemy_ibmi/base.py]: I opened https://github.com/IBM/sqlalchemy-ibmi/issues/41 to tackle that.

File: sqlalchemy_ibmi/base.py (hunk 2 above)
Comment (kadler) [reply sqlalchemy_ibmi/base.py]: Ok, sounds good. Personally would have rather seen this work done in a separate PR.

File: sqlalchemy_ibmi/base.py (hunk 3 above)
Comment (kadler) [reply sqlalchemy_ibmi/base.py]: No clue, but I'm not sure _why_ it was added or what it would be used for so I'm inclined to just leave it be gone.

File: sqlalchemy_ibmi/base.py [hunk 5]
```diff
@@ -636,19 +637,41 @@ def get_rowcount(self):
     def fire_sequence(self, seq, type_):
         return self._execute_scalar(
             "SELECT NEXTVAL FOR " +
-            self.connection.dialect.identifier_preparer.format_sequence(seq) +
+            self.connection.dialect.preparer.format_sequence(seq) +
```
Comment (kadler) [inline sqlalchemy_ibmi/base.py]: I think this should be left as identifier_preparer.

dialect here should be our dialect, which subclasses DefaultDialect which sets up `identifier_preparer` here: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/default.py#L249, while its `preparer` is actually a class not an instance: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/default.py#L48 and we have the same in our dialect: https://github.com/IBM/sqlalchemy-ibmi/blob/master/sqlalchemy_ibmi/base.py#L677

I can understand the confusion because in the SQLCompiler class, which Db2Compiler subclasses, it sets `preparer` to be the dialect's `identifier_preparer`: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/sql/compiler.py#L411

File: sqlalchemy_ibmi/base.py [hunk 6]
```diff
@@ -813,7 +804,7 @@ def _get_default_schema_name(self, connection):
 
     @property
     def default_schema_name(self):
-        return self.dialect.default_schema_name
+        return self._get_default_schema_name()
```
Comment (kadler) [inline sqlalchemy_ibmi/base.py]: _get_default_schema_name takes a connection object, but we're not passing it. Probably need to pass `self.connection`.

File: sqlalchemy_ibmi/base.py [hunk 7]
```diff
@@ -802,10 +802,6 @@ def _get_default_schema_name(self, connection):
             default_schema_name = default_schema_name.strip()
         return self.normalize_name(default_schema_name)
 
-    @property
-    def default_schema_name(self):
-        return self._get_default_schema_name()
```
Comment (kadler) [inline sqlalchemy_ibmi/base.py]: I didn't think removing this would work, but it looks like the DefaultDialect will call `_get_default_schema_name` during `initialize`: https://github.com/zzzeek/sqlalchemy/blob/master/lib/sqlalchemy/engine/default.py#L311-L314 Awesome, though it probably should have been done in a separate PR to explain why.
